src/playback-watcher.js

```diff
@@ -156,20 +167,63 @@ export default class PlaybackWatcher {
   }
 
   /**
-   * Handler for situations when we determine the player is waiting
+   * Handler for situations when we determine the player is waiting.
    *
    * @private
    */
   waiting_() {
+    if (this.techWaiting_()) {
+      return;
+    }
+
+    // All tech waiting checks failed. Use last resort correction
+    let currentTime = this.tech_.currentTime();
+    let buffered = this.tech_.buffered();
+    let currentRange = Ranges.findRange(buffered, currentTime);
+
+    // Sometimes the player can stall for unknown reasons within a contiguous buffered
+    // region with no indication that anything is amiss (seen in Firefox). Seeking to
+    // currentTime is usually enough to kickstart the player. This checks that the player
+    // is currently within a buffered region and there is at least half a second
+    // of forward buffer so that this isn't triggered when the player is just buffering
+    // due to slow connection.
+    // Note: This is not done when the `waiting` event fired by the tech because the tech
+    // also fires `waiting` when the player is buffering in low bandwidth scenarios, which
+    // requires no action from playback watcher.
+    if (currentRange.length && currentTime <= currentRange.end(0) - 0.5) {
```

Comment:
I'm not sure how necessary the 0.5 safety offset is now that in `checkCurrentTime_` we check to see if we are at the end of the buffer. Thoughts?